admin2: refactor ban functionality (resolves #151)

jlillis opened the pull request:
To satisfy #151

This PR redesigns the add ban widget. It can be accessed by selecting a player and clicking "Ban" in the main tab, or clicking "Add ban" from the player tab. When accessed from the main tab it will autofill the target player's IP and serial. You can chose whether to add either the IP or serial (or both) to the ban.

![image](https://github.com/multitheftauto/mtasa-resources/assets/2784366/fae66166-c9b4-4744-85b6-e583c9d43188)

Some remaining tasks:

- [x] Add ability to define custom duration
- [x] Mask player identifiers when banning from players tab with "streamer mode" on
- [x] Agree upon and set new default ban durations
- [x] Consider refactoring server code to resolve misc. errors
- [x] Add server log and chatbox outputs for "offline" bans
- [x] Fix server errors when using `banip` and `banserial` commands: 
```
[01:08:26] ERROR: [mtasa-resources]\[admin]\admin2\admin_coroutines.lua:38: ...sa-resources]\[admin]\admin2\server\admin_server.lua:410: attempt to call global 'BanIP' (a nil value) [string "?"]
[01:08:26] INFO: ...sa-resources]\[admin]\admin2\server\admin_server.lua:410: attempt to call global 'BanIP' (a nil value)
[01:08:30] ERROR: [mtasa-resources]\[admin]\admin2\admin_coroutines.lua:38: ...sa-resources]\[admin]\admin2\server\admin_server.lua:415: attempt to call global 'isValidSerial' (a nil value) [string "?"]
[01:08:30] INFO: ...sa-resources]\[admin]\admin2\server\admin_server.lua:415: attempt to call global 'isValidSerial' (a nil value)
[01:08:33] ADMIN: STEELMEAT has banned $player ()
[01:08:33] WARNING: [mtasa-resources]\[admin]\admin2\server\admin_functions.lua:32: Bad argument @ 'banPlayer' [Expected player at argument 1, got nil]
```
- [x] Fix unban button and command
- [x] Re-implement "ban details" widget
- [x] Remove references to ban (community) username

~~While there are some remaining tasks I consider this ready for testing.~~

I haven't been able to test this with multiple players so I don't know if everything is working properly, especially the sync of added/deleted bans in the bans tab.

Dark-Dragon commented:
Here are a couple of things I ran into while testing (all except the very last are manual bans from the ban tab, not the players tab):

- I keep being told my serial is invalid when trying to ban my own serial
- After receiving an error for not providing an IP/Serial or an invalid one the the confirmation message box still popped up
- Sometimes I had to reselect a ban duration when submitting multiple bans in a row (can't quite tell how to reproduce just yet)
- Confirmation message box pops up despite not checking IP or Serial
- Maybe the confirmation message box type should be question rather than warning?
- After confirming a ban there's a lack of feedback. I'm not sure if the gridlist is supposed to update immediately, but if it is, then it didn't appear to work during my testing. I believe normally there should be a chat output too, shouldn't there?
- The 1 second ban I presume is just used for testing
- Maybe the confirmation could say the ban duration again?
- Generally the widgets in admin2 appear to leave less empty padding space around the sides than this one, might be worth considering to adjust for consistency?
- Hide sensitive player data should probably mask the IP/Serial if checked when banning through players tab.

jlillis commented:
>     * I keep being told my serial is invalid when trying to ban my own serial
Fixed.

>     * After receiving an error for not providing an IP/Serial or an invalid one the the confirmation message box still popped up
Fixed.

>     * Sometimes I had to reselect a ban duration when submitting multiple bans in a row (can't quite tell how to reproduce just yet)
Added a line to reset the ban duration to default when the widget is shown, which should fix this.

>     * Confirmation message box pops up despite not checking IP or Serial
Fixed.

>     * Maybe the confirmation message box type should be question rather than warning?
I think it's more of a question rather than a warning but if anyone else feels otherwise it's easy to change.

>     * After confirming a ban there's a lack of feedback. I'm not sure if the gridlist is supposed to update immediately, but if it is, then it didn't appear to work during my testing. I believe normally there should be a chat output too, shouldn't there?
Partially fixed. I had to dig around the serverside code some more to figure out how the ban list sync was working. New bans should appear in the ban list as soon as they are added. It's not outputting to server log/chatbox currently, added that to my todo list.

>     * The 1 second ban I presume is just used for testing
Correct.

>     * Maybe the confirmation could say the ban duration again?
There isn't a ton of room in those messagebox widgets, but I'll play with them and see what works.

>     * Generally the widgets in admin2 appear to leave less empty padding space around the sides than this one, might be worth considering to adjust for consistency?
Just reduced the horizontal padding by 50%.

>     * Hide sensitive player data should probably mask the IP/Serial if checked when banning through players tab.
Added to todo list.

jlillis commented:
I've finally got around to crossing off the rest of my todo list and am calling this done and ready for more testing. I "fixed" the unimplemented ban/unban commands by just removing them - if there are enough complaints I'll go ahead and get them added back in.
